sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpProducer.cs

```diff
@@ -199,13 +199,59 @@ internal class AmqpProducer : TransportProducer
                                                                               CancellationToken cancellationToken)
         {
             Argument.AssertNotNull(options, nameof(options));
+            Argument.AssertNotClosed(_closed, nameof(AmqpProducer));
+
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
 
             // Ensure that maximum message size has been determined; this depends on the underlying
             // AMQP link, so if not set, requesting the link will ensure that it is populated.
 
             if (!MaximumMessageSize.HasValue)
             {
-                await SendLink.GetOrCreateAsync(RetryPolicy.CalculateTryTimeout(0)).ConfigureAwait(false);
+                var failedAttemptCount = 0;
+                var retryDelay = default(TimeSpan?);
+                var tryTimeout = RetryPolicy.CalculateTryTimeout(0);
+
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    try
```

Comment:
I think we may have hit the point where we've got this logic duplicated in enough places where it makes sense to abstract it.   Taking a note for myself to include that as part of the processor implementation that is underway.

For this PR, let's leave it inline unless you're motivated to do otherwise - in which case, let's sync.

Reply:
I'll merge this PR for now. We can get this done in another one.